source/common/upstream/load_balancer_impl.cc

```diff
@@ -473,9 +477,9 @@ void ZoneAwareLoadBalancerBase::regenerateLocalityRoutingStructures() {
     return;
   }
   HostSet& host_set = *priority_set_.hostSetsPerPriority()[priority];
-  ASSERT(host_set.healthyHostsPerLocality().hasLocalLocality());
-  const size_t num_localities = host_set.healthyHostsPerLocality().get().size();
-  ASSERT(num_localities > 0);
+  const HostsPerLocality& upstreamHostsPerLocality = host_set.healthyHostsPerLocality();
+  const size_t num_upstream_localities = upstreamHostsPerLocality.get().size();
+  ASSERT(num_upstream_localities >= 2);
```

Comment:
This change is incredibly risky. What concerns me here is that a lot of code is changing independent of whether the feature flag is set or not. Even if it means code duplication, is there any way to split the changes out so that when the feature flag is not set there are no changes? This would make me feel more comfortable that the change can be rolled back if there is an issue. If there are no issues we can clean up the duplication later.

Reply:
Do you have an example of another PR/change that does this? I'm mostly looking for design patterns/best practices/etc

I agree this is a good idea

Reply:
I can't recall one recently, but I know we have done things like this in the past for super risky changes. cc @alyssawilk for thoughts.

I think the regression risk of this change is very high (so I would change that in the PR description.). This is code that has to my knowledge not changed very much since almost the very beginning of the project, and will likely effect many users. To be clear, I'm very happy you are fixing this, I just want to be careful. :)

/wait

Reply:
For high risk changes like this where you have lots of invasive bits, sometimes folks temporarily clone the entire class, but that'd suck given the inheritance model.
My best thought is clone functions which are invasively changed, latch runtime status in the base class and do things like
if (legacy_mode) {
calculateLocalityPercentagesLegacy() <- old function here
} else {
calculateLocalityPercentages() <- your new function
}

actually I think if you clone the calculateLocality and regenerateStructures functions
you'd only need if(legacy_mode)  in a few places.

Reply:
that makes sense, thanks for the info

most of this stuff hasn't been touched in ~5 years at least, so it is definitely higher risk

Reply:
I'll work on cloning stuff

Reply:
I've added a new runtime guard `envoy.reloadable_features.locality_routing_use_new_routing_logic`. Not super happy with the name, lmk if you have better suggestions.

I've split all the changed methods into legacy/current clones, where the legacy versions are completely unchanged. We latch the runtime state on construction, and use the legacy version if the runtime flag is disabled.

The only method I didn't guard is `ZoneAwareLoadBalancerBase::tryChooseLocalLocalityHosts`, because it's completely separate from the rest of the stuff I've guarded and the changes are extremely minor and entirely compatible with both versions. If you'd like me to include this in the guard, it should be pretty simple.

I've changed the `load_balancer_impl_tests` to include a parameter for the legacy vs current flag so that all the tests run using both code paths. For the few tests where behavior changes depending on which path is used, I test behavior in both cases depending on what the parameter value is.